### DISCUSSION TITLE
How to Create a Data History

### DISCUSSION THREAD

**Discussion by thibault-chausson** [supabase/supabase#19724] [Questions]
Hello,
I am currently developing a web application and using SupaBase as my database. Each person can write a note, and I would like to keep a version of each note. This versioning would allow me to offer my users a history of their modifications.
How can I do this with SupaBase? I have seen solutions like pgaudit, is this the way to go? Or would it be better to manually create a "history" table and use triggers and functions to duplicate the data?
Thanks in advance for your help!

**Comment by jonathanlal** [comment]
also interested.... apparently there's an extension called pg_temporal but i don't find any docs on it

**Comment by arjunlol** [comment]
There's a few different ways to implement this depending on your requirements. PGAudit sends detailed audit logs to the standard PostgreSQL output logs, but doesn't record events to a table. [This guide](https://blog.bemi.io/the-ultimate-guide-to-postgresql-data-change-tracking/) is a good reference of the pros/cons of all the approaches.

**Comment by romanr** [comment]
Every solution requires creating low-level SQL queries and implement object versioning and history retrieval manually.

I was hoping there's open source wrapper that performs all of that, something like https://bemi.io/.

Supabase has `supa_audit` module, no examples how to connect it all to achieve object records versioning.
